src/Model.php

```diff
@@ -1862,21 +1803,60 @@ public function atomic(\Closure $fx, Persistence $persistence = null)
 
     // }}}
 
-    // {{{ Support for actions
+    // {{{ Support for query and expressions
 
     /**
-     * Execute action.
-     *
-     * @param string $mode
-     * @param array  $args
-     *
-     * @return Query
+     * @deprecated use toQuery instead - will be removed in dec-2020
      */
-    public function action($mode, $args = [])
+    public function action($mode, $args = []): Persistence\AbstractQuery
     {
-        $this->checkPersistence('action');
+        'trigger_error'('Method Model::action is deprecated. Use Model::query instead', E_USER_DEPRECATED);
+
+        $this->checkPersistence('query');
+
+        // start backward compatibility -->
+        switch ($mode) {
+            case 'fx':
+            case 'fx0':
+                [$fx, $field] = $args;
+                $alias = $args['alias'] ?? null;
+                $coalesce = $mode === 'fx0';
+
+                $args = [$fx, $field, $alias, $coalesce];
+                $mode = 'aggregate';
+
+                break;
+            case 'field':
+                $args = [$args[0], $args['alias'] ?? null];
+
+                break;
+            case 'count':
+                $args = [$args['alias'] ?? null];
 
-        return $this->persistence->action($this, $mode, $args);
+                break;
+            default:
+                break;
+        }
+        // <-- end backward compatibility
+
+        $query = $this->persistence->query($this);
+
+        if (!method_exists($query, $mode)) {
+            throw (new Exception('Unsupported query mode'))
+                ->addMoreInfo('type', $mode);
+        }
+
+        return $this->persistence->query($this)->{$mode}(...$args);
+    }
+
+    /**
+     * Get query object to perform query on raw persistence data.
+     */
+    public function query(): Persistence\AbstractQuery
```

Comment:
the issue is that it does not return query, but instead it creates a new one, thus toQuery :)

Reply:
`Model::query` changed to `Model::toQuery`